### ISSUE TITLE
{Strimzi] MirrorMaker2 Integration

### ISSUE THREAD

**Issue by viscory** [cricket007/schema-registry-transfer-smt#38]
HI

To my understanding, Strimzi's MM2 implementation does not allow for custom images to be used. Is there a way to integrate MirrorMaker2 with your SMT then? 

Thank you for your work!

**Comment by OneCricketeer** [comment]
It does allow custom images

https://github.com/strimzi/strimzi-kafka-operator/blob/main/helm-charts/helm3/strimzi-kafka-operator/crds/045-Crd-kafkamirrormaker.yaml#L66

**Comment by viscory** [comment]
Hi

My bad, it does allow for custom images. I ran mirrormaker2 with a custom docker image which had your smt transplanted in

MirrorMaker2 does copy all of the data into the target cluster. However, the schemas are not registered into apicurio. This is the error we get: 
\
  

> Caused by: org.apache.kafka.common.errors.SerializationException: Unexpected byte[] length 3 for Avro record value.
> 	  at cricket.jmoore.kafka.connect.transforms.SchemaRegistryTransfer.apply(SchemaRegistryTransfer.java:176)
> 	  at org.apache.kafka.connect.runtime.TransformationStage.apply(TransformationStage.java:57)
> 	  at org.apache.kafka.connect.runtime.TransformationChain.lambda$apply$0(TransformationChain.java:53)
> 	  at org.apache.kafka.connect.runtime.errors.RetryWithToleranceOperator.execAndRetry(RetryWithToleranceOperator.java:180)
> 	  at org.apache.kafka.connect.runtime.errors.RetryWithToleranceOperator.execAndHandleError(RetryWithToleranceOperator.java:214)
> 	  ... 13 more

	
The mirrormaker2 config we are using:

> 
    apiVersion: kafka.strimzi.io/v1beta2
    kind: KafkaMirrorMaker2
    metadata:
      name: mm2
    spec:
      image: example-registry/strimzi/kafka-connector:0.36.1-kafka-3.5.1_1
      clusters:
        - alias: source
          bootstrapServers: 'example-src-cluster:9092'
          config:
            key.converter.schema.registry.url: 'http://example-cp-registry:8081'
            value.converter.schema.registry.url: 'http://example-cp-registry:8081'
        - alias: target
          bootstrapServers: 'example-target-cluster:9092'
          config:
            config.storage.replication.factor: -1
            offset.storage.replication.factor: -1
            status.storage.replication.factor: -1
            key.converter.schema.registry.url: 'http://example-apicurio-registry:8081'
            value.converter.schema.registry.url: 'http://example-apicurio-registry:8081'
      connectCluster: target
      mirrors:
        - sourceCluster: source
          sourceConnector:
            config:
              replication.policy.class: io.strimzi.kafka.connect.mirror.IdentityReplicationPolicy
              replication.policy.separator: ''
              sync.topic.acls.enabled: 'false'
              transforms: AvroSchemaTransfer
              transforms.AvroSchemaTransfer.type: "cricket.jmoore.kafka.connect.transforms.SchemaRegistryTransfer"
              transforms.AvroSchemaTransfer.src.schema.registry.url: 'http://example-cp-registry:8081'
              transforms.AvroSchemaTransfer.dest.schema.registry.url: 'http://example-apicurio-registry:8080/apis/ccompat/v6'
            tasksMax: 10
          targetCluster: target
          topicsPattern: .*
      replicas: 1
      version: 3.5.1*

**Comment by viscory** [comment]
Update, changed some configs (added bytearrayconverter as the key/value converters and got this error

```
Caused by: org.apache.kafka.common.errors.SerializationException: Unknown magic byte!38
at cricket.jmoore.kafka.connect.transforms.SchemaRegistryTransfer.copySchema(SchemaRegistryTransfer.java:235)
```

I'd appreciate any help I get from you on this!

**Comment by OneCricketeer** [comment]
This SMT only supports the Confluent wire format documented at - https://docs.confluent.io/platform/current/schema-registry/fundamentals/serdes-develop/index.html#wire-format

As written there, the magic byte must always be zero. 

I do not use Apicurio, so am not sure how they're records would be encoded, but from what I understand, they do have a different format, and a different api endpoint to offer Confluent compatibility

**Comment by OneCricketeer** [comment]
Looking back at Apicurio code, [it also uses 0x0 as the magic byte](https://github.com/Apicurio/apicurio-registry/blob/main/serdes/serde-common/src/main/java/io/apicurio/registry/serde/AbstractKafkaSerDe.java#L122), so that tells me that some producer process (possibly upstream of the mirroring) is sending incorrectly serialized data. 

This also matches the conditions of the previous error, since Avro encoded Registry data must be at least 6 bytes in length 

> `Unexpected byte[] length 3`

More specifically, if you let `kafka-avro-console-consumer --from-beginning` (or the equivalent Apicurio command, if one exists) run to completion, then you would expect the same error. 

In other words, Kafka broker isn't going to prevent bad client configuration. This is why Confluent offers enterprise feature of [Broker Schema ID validation](https://docs.confluent.io/platform/current/schema-registry/schema-validation.html) on their Cloud or on-prem Confluent Server offering (i.e. not an option for Strimzi without some additional proxy/validation layer)
